once conflict review feature is unmerged w/ undo reintroduce merge button

jackgrossman18 opened the pull request:
ref #1635

Commit:
once conflict review feature is unmerged w/ undo reintroduce merge button
    ref #1635

maxgrossman commented:
This worked for sure when I clicked merge, then undo back to before all the actions within a merge (change tags, delete).

However, if I use undo/redo to go to the head and base of the undo/redo stack (click undo or redo until I cannot anymore in either direction), the button gets all out of sync.

Perhaps we can listen to the undone/redone events fired by the history class?
The class is accessible on context so it'd be avail in this class.

I think you'd be able to say something like...

on `undone`, if the thing that is being undone (via the annotation on the graph) is changed tags, we know we can show the merge button. This is because the first step of the merge is to change the tags of the target feature.

on `redone` If the thing being redone is that then you'd be able to say ok hide the merge button. (since that's the inverse of what we care about for undone)

maxgrossman reviewed (changes requested):
Nice! I added a comment. This seems to work in one direction (when I click undone).
We'll need this to be bi-directional (undoing and redoing), and I added an idea of how we might be able to achieve it.

brianhatchl committed:
Merge remote-tracking branch 'origin/hoot2x' into 1635

brianhatchl committed:
Merge remote-tracking branch 'origin/hoot2x' into 1635

brianhatchl committed:
move merge button check to merge.js and use history listener

brianhatchl committed:
minimize diff

brianhatchl committed:
merge.js has its own access to these params

brianhatchl committed:
refactor quite a bit

use feature types not review type to determine isPoiPoly
add flash warning that poi->poly can't be reversed
use mergeArray from/to when merging instaed of currentFeatures

brianhatchl committed:
lint fix

brianhatchl commented:
@bobi7 I made the changes suggested by @maxgrossman.  Rather than checking for what change the unde/redo is peforming, it seems sufficient to just check the review features and make sure they are eligible for merge.  The test at play here is that both review features exist in the graph, which is not true after merge, but becomes true again when the merge is undone.

brianhatchl committed:
one review feature may be null so check this first

brianhatchl committed:
alternate implementation for #1480 update tag table after merge

brianhatchl committed:
lint fix

brianhatchl commented:
should also fix #1006 #1480

brianhatchl commented:
might also add a fix for #1476

brianhatchl commented on modules/Hoot/ui/conflicts/info.js:
```diff
@@ -26,19 +27,35 @@ export default class ConflictMetadata {
     buildTagTable() {
-        let colData    = this.data.currentFeatures,
-            tags1      = this.filterTags( colData[ 0 ] ? colData[ 0 ].tags : {} ),
+        let currentRelation;
+        try {
+            currentRelation = this.instance.graphSync.getCurrentRelation();
+        } catch (e) {
+            //current relation not found which is ok
+        }
+        //if triggered by history change and not in review mode return
+        if (!currentRelation) return;
```
@jackgrossman18 because this method listens to history changes, it will be called outside the conflation review context and so getting the current relation may throw an "entity not found" error.

brianhatchl commented on modules/Hoot/ui/conflicts/info.js:
```diff
@@ -26,19 +27,35 @@ export default class ConflictMetadata {
+        //if triggered by history change and not in review mode return
+        if (!currentRelation) return;
+
+        let features = this.data.currentFeatures,
+            feature = Hoot.context.hasEntity( features[0].id ),
+            againstFeature = Hoot.context.hasEntity( features[1].id );
+
+        let colData = [feature, againstFeature];
```
@jackgrossman18 this seems to me a simpler way to handle if a feature has been merged and 1) had tags updated or 2) been deleted.  By grabbing the features from the graph instead of the stale `currentFeatures` the tag table will now update correctly anytime review merges happen.

brianhatchl commented on modules/Hoot/ui/conflicts/map.js:
```diff
@@ -58,17 +58,7 @@ export default class Map {
-        } else {
-            this.instance.merge.toggleMergeButton( true );
-        }
+        this.instance.merge.checkMergeButton();
```
@jackgrossman18 move this logic inside `merge.js`

brianhatchl commented on modules/Hoot/ui/conflicts/merge.js:
```diff
@@ -23,13 +25,25 @@ export default class Merge {
+        this.isPoiPoly = false;
+
+        let that = this;
+        Hoot.context.history()
+            .on('undone.review_merge', function() {
+                that.checkMergeButton.call(that);
+            });
+        Hoot.context.history()
+            .on('redone.review_merge', function() {
+                that.checkMergeButton.call(that);
+            });
```
@jackgrossman18 listen for undo/redo and call the same function that is called when a new review is loaded.

brianhatchl commented on modules/Hoot/ui/conflicts/merge.js:
```diff
@@ -38,28 +52,20 @@ export default class Merge {
-            featureUpdate,
-            featureDelete,
+        let reverse  = d3.event.ctrlKey || d3.event.metaKey && !this.isPoiPoly,
+            featureUpdate = _cloneDeep((reverse) ? this.mergeArrow.from : this.mergeArrow.to),
+            featureDelete = _cloneDeep((reverse) ? this.mergeArrow.to : this.mergeArrow.from),
```
@jackgrossman18 we have to deep clone here because we modify the feature tags on line 68.  If we do a merge, then undo it and reverse it the element merge service errors because both featuers have the `hoot:merge:target=yes` tag.

brianhatchl replied:
Also when checking if reverse merge is valid, we disallow if it is POI->Poly.  We can simply disallow because code in `checkMergeButton` ;omr 358 makes sure the poi and poly features are always in the right order.

brianhatchl commented on modules/Hoot/ui/conflicts/merge.js:
```diff
@@ -328,11 +314,23 @@ export default class Merge {
 
-        let pt1   = d3.geoCentroid( this.mergeArrow.to.asGeoJSON( Hoot.context.graph() ) ),
-            pt2   = d3.geoCentroid( this.mergeArrow.from.asGeoJSON( Hoot.context.graph() ) ),
+        let pt1   = d3.geoCentroid( this.mergeArrow.from.asGeoJSON( Hoot.context.graph() ) ),
+            pt2   = d3.geoCentroid( this.mergeArrow.to.asGeoJSON( Hoot.context.graph() ) ),
```
I found that the original code was putting the arrow direction in the opposite from the intended from/to direction

brianhatchl commented on modules/Hoot/ui/conflicts/merge.js:
```diff
@@ -342,4 +340,34 @@ export default class Merge {
+
+    checkMergeButton() {
+        let features = _clone( this.data.currentFeatures ),
+            toFeature = Hoot.context.hasEntity( features[0].id ),
+            fromFeature = Hoot.context.hasEntity( features[1].id ),
```
Use dynamic features retrieved from graph.  This allows the show/hide merge button to work on undo/redo.

brianhatchl commented on modules/Hoot/ui/conflicts/merge.js:
```diff
@@ -342,4 +340,34 @@ export default class Merge {
+            relation = this.instance.graphSync.getCurrentRelation();
+
+        if ( toFeature && fromFeature) {
+            this.isPoiPoly = (toFeature.type === 'node' && fromFeature.type === 'way')
+                || (toFeature.type === 'way' && fromFeature.type === 'node');
```
We need to define POI->Poly merge this way (instead of using `hoot:review:type` because after a merge, other review relations will be updated to refer to the merged element.  After a poi->poly merge the referred element may no long be a POI and may be two polys but the review type remains unchanged.  This may change when we implement this long ago ticket #1003 to allow poly->poly merges.

brianhatchl commented on modules/Hoot/ui/conflicts/merge.js:
```diff
@@ -342,4 +340,34 @@ export default class Merge {
+
+            if (this.isPoiPoly || isPoiPoi) {
+                this.toggleMergeButton( false );
+                if (this.isPoiPoly) {
+                    let poi = (toFeature.type === 'node') ? toFeature : fromFeature;
+                    let poly = (toFeature.type === 'way') ? toFeature : fromFeature;
+                    this.activateMergeArrow( poi, poly ); //always merge poi->poly
```
Make sure the feature order for the merge array is always poi->poly

brianhatchl committed:
restore merge button on failed merge

fixes #1476

bobi7 commented on modules/Hoot/ui/conflicts/merge.js:
```diff
@@ -38,28 +52,20 @@ export default class Merge {
-            reverse  = d3.event.ctrlKey || d3.event.metaKey || this.mergeArrow.reverseFeatures,
-            featureUpdate,
-            featureDelete,
+        let reverse  = d3.event.ctrlKey || d3.event.metaKey && !this.isPoiPoly,
```
should we wrap the `d3.event.ctrlKey || d3.event.metaKey` in parenthesis for readability?

brianhatchl committed:
readability

brianhatchl committed:
display element merge server error if available

refs https://github.com/ngageoint/hootenanny/pull/4176

brianhatchl committed:
remove code that makes merge fail for testing purposes

brianhatchl commented:
Also fixes #1214

brianhatchl commented:
and fixes #1795
